compiler/rustc_parse/src/parser/pat.rs

```diff
@@ -915,6 +912,61 @@ impl<'a> Parser<'a> {
         Ok(PatKind::TupleStruct(qself, path, fields))
     }
 
+    /// Parses `box pat`
+    fn parse_pat_box(&mut self) -> PResult<'a, PatKind> {
+        let box_kw = self.token.ident().unwrap().0;
+        self.bump();
+
+        // All of these can follow identifiers in patterns, but cannot be at
+        // the beginning of a pattern. If we see them, the user probably meant
+        // `box` to be an identifier.
+        // `(` and `::` also can follow identifiers, but they can also follow
+        // `box` (`box (tuple)`, `box ::path`) and thus we can't suggest anything
+        // about them unless we try to completely parse a pattern. Luckily,
+        // if the use meant them to be an identifier (a tuple struct or a path,
+        // respectively) we would already suggest escaping at its declaration site.
+        // Another one is `|` (for or pattern), but or-ing an identifier is pretty
+        // redundant since it always match.
+        let ident_follow = &[
+            token::Eq,
+            token::Colon,
+            token::Comma,
+            token::Semi,
+            token::At,
+            token::OpenDelim(Delimiter::Brace),
+            token::CloseDelim(Delimiter::Brace),
+            token::CloseDelim(Delimiter::Parenthesis),
+        ];
```

Comment:
I feel really uncomfortable listing them all here because it really isn't future-proof. But I have found no better way.

Reply:
Haven't thought about that, however this doesn't include (nor I checked) all tokens that can't start a pattern, only those that also can follow an ident. It does make sense to centralize that, however.

Reply:
Sure, it doesn't need to be an exhaustive list, just a good heuristic for "definitely not a pattern start" in a case like this.

Reply:
What about tokens like `=`, that can't open a pattern but are also meaningless in this context? Or the criteria is "can't open a pattern but in the middle of a pattern"?

Reply:
maybe this needs a more creative name than `isnt_pattern_start`, but i would like if it got moved to a helper, regardless.

Reply:
Oops, `=` is actually on the list :) There are other tokens however that are not. I think `isnt_pattern_start()` is pretty good, as long as I put it in `pat.rs`. I can leave a comment explaining this is not complete and another tokens can be added as necessary in the future.